File: java-workspace2/src/백준/다이나믹프로그래밍/일이삼더하기.java
package 백준.다이나믹프로그래밍;

import java.util.Scanner;

public class 일이삼더하기 {
	static Scanner scanner = new Scanner(System.in);
	static int testcase;
	static int answer;

	public static void main(String[] args) {
		testcase = scanner.nextInt();
		for (int i = 0; i < testcase; i++) {
			answer = 0;
			search(0, scanner.nextInt());
			System.out.println(answer);
		}
	}

	private static void search(int sum, int target) {
		if (sum == target) {
			answer++;
			return;
		}
		if (sum > target) {
			return;
		}
		for (int i = 1; i <= 3; i++) {
			search(sum + i, target);
		}
	}
}

File: python-workspace/dfs/일이삼더하기.py
import sys


def back_tracking(units, sum, target):
    global answer
    if sum == target:
        answer += 1
    for unit in units:
        if sum + unit <= target:
            back_tracking(units, sum + unit, target)


answer = 0
testcase = int(sys.stdin.readline())
for _ in range(testcase):
    target = int(sys.stdin.readline())
    units = [1, 2, 3]
    answer = 0
    back_tracking(units, 0, target)
    print(answer)
